transactron/lib/allocators.py

```diff
@@ -60,3 +61,74 @@ def _(_, ident):
             m.d.sync += not_used.bit_select(ident, 1).eq(1)
 
         return m
+
+
+class PreservedOrderAllocator(Elaboratable):
+    """Allocator with allocation order information.
+
+    This module allows to allocate and deallocate identifiers from a
+    continuous range. The order of allocations is preserved in the form of
+    a permutation of identifiers. Smaller positions correspond to earlier
+    (older) allocations.
+
+    Attributes
+    ----------
+    alloc : Method
+        Allocates a fresh identifier.
+    free : Method
+        Frees a previously allocated identifier.
+    free_idx : Method
+        Frees a previously allocated identifier at the given index of the
+        allocation order.
+    order : Method
+        Returns the allocation order as a permutation of identifiers
+        and the number of allocated identifiers.
+    """
+
+    def __init__(self, entries: int):
+        self.entries = entries
+
+        self.alloc = Method(o=[("ident", range(entries))])
+        self.free = Method(i=[("ident", range(entries))])
+        self.free_idx = Method(i=[("idx", range(entries))])
+        self.order = Method(
+            o=[("used", range(entries + 1)), ("order", ArrayLayout(range(self.entries), self.entries))],
+            nonexclusive=True,
+        )
+
+    def elaborate(self, platform) -> TModule:
+        m = TModule()
+
+        order = Signal(ArrayLayout(range(self.entries), self.entries), init=list(range(self.entries)))
+        used = Signal(range(self.entries + 1))
+        incr_used = Signal(range(self.entries + 1))
+
+        m.d.comb += incr_used.eq(used + self.alloc.run)
+        m.d.sync += used.eq(incr_used - self.free_idx.run)
+
+        @def_method(m, self.alloc, ready=used != self.entries)
+        def _():
+            return {"ident": order[used]}
+
+        @def_method(m, self.free_idx)
+        def _(idx):
+            for i in range(self.entries - 1):
+                with m.If((i >= idx) & (i + 1 < incr_used)):
+                    m.d.sync += order[i].eq(order[i + 1])
+            for i in range(self.entries):
+                with m.If(i + 1 == incr_used):
+                    m.d.sync += order[i].eq(order[idx])
```

Comment:
Not very critical, but I think it can be cheaper in hardware by always shifting `[idx:]` suffix and putting freed identifier at end of `order` (avoids second mux). Additionally it would produce more balanced identifier use.
Would it break some intended use cases, or is it worth adding?

Reply:
As I consider free IDs to be undistinguishable, the solution you present is indeed better. Done.